File: Catalog core/src/main/java/cz/vhromada/catalog/facade/impl/MovieFacadeImpl.java
package cz.vhromada.catalog.facade.impl;

import java.util.ArrayList;
import java.util.List;

import cz.vhromada.catalog.common.Time;
import cz.vhromada.catalog.domain.Genre;
import cz.vhromada.catalog.domain.Medium;
import cz.vhromada.catalog.domain.Movie;
import cz.vhromada.catalog.entity.GenreTO;
import cz.vhromada.catalog.entity.MediumTO;
import cz.vhromada.catalog.entity.MovieTO;
import cz.vhromada.catalog.facade.MovieFacade;
import cz.vhromada.catalog.service.CatalogService;
import cz.vhromada.catalog.validators.MovieTOValidator;
import cz.vhromada.converters.Converter;
import cz.vhromada.validators.Validators;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

/**
 * A class represents implementation of facade for movies.
 *
 * @author Vladimir Hromada
 */
@Component("movieFacade")
@Transactional
public class MovieFacadeImpl implements MovieFacade {

    /**
     * TO for movie argument
     */
    private static final String MOVIE_TO_ARGUMENT = "TO for movie";

    /**
     * TO for genre argument
     */
    private static final String GENRE_TO_ARGUMENT = "TO for genre";

    /**
     * Service for movies
     */
    private CatalogService<Movie> movieService;

    /**
     * Service for genres
     */
    private CatalogService<Genre> genreService;

    /**
     * Converter
     */
    private Converter converter;

    /**
     * Validator for TO for movie
     */
    private MovieTOValidator movieTOValidator;

    /**
     * Creates a new instance of MovieFacadeImpl.
     *
     * @param movieService     service for movies
     * @param genreService     service for genres
     * @param converter        converter
     * @param movieTOValidator validator for TO for movie
     * @throws IllegalArgumentException if service for movies is null
     *                                  or service for genres is null
     *                                  or converter is null
     *                                  or validator for TO for movie is null
     */
    @Autowired
    public MovieFacadeImpl(final CatalogService<Movie> movieService,
            final CatalogService<Genre> genreService,
            @Qualifier("catalogDozerConverter") final Converter converter,
            final MovieTOValidator movieTOValidator) {
        Validators.validateArgumentNotNull(movieService, "Service for movies");
        Validators.validateArgumentNotNull(genreService, "Service for genres");
        Validators.validateArgumentNotNull(converter, "Converter");
        Validators.validateArgumentNotNull(movieTOValidator, "Validator for TO for movie");

        this.movieService = movieService;
        this.genreService = genreService;
        this.converter = converter;
        this.movieTOValidator = movieTOValidator;
    }

    @Override
    public void newData() {
        movieService.newData();
    }

    @Override
    public List<MovieTO> getMovies() {
        return converter.convertCollection(movieService.getAll(), MovieTO.class);
    }

    /**
     * @throws IllegalArgumentException {@inheritDoc}
     */
    @Override
    public MovieTO getMovie(final Integer id) {
        Validators.validateArgumentNotNull(id, "ID");

        return converter.convert(movieService.get(id), MovieTO.class);
    }

    /**
     * @throws IllegalArgumentException                              {@inheritDoc}
     * @throws cz.vhromada.validators.exceptions.ValidationException {@inheritDoc}
     */
    @Override
    public void add(final MovieTO movie) {
        movieTOValidator.validateNewMovieTO(movie);
        for (final GenreTO genre : movie.getGenres()) {
            Validators.validateExists(genreService.get(genre.getId()), GENRE_TO_ARGUMENT);
        }

        movieService.add(converter.convert(movie, Movie.class));
    }

    /**
     * @throws IllegalArgumentException                                  {@inheritDoc}
     * @throws cz.vhromada.validators.exceptions.ValidationException     {@inheritDoc}
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException {@inheritDoc}
     */
    @Override
    public void update(final MovieTO movie) {
        movieTOValidator.validateExistingMovieTO(movie);
        final Movie movieEntity = movieService.get(movie.getId());
        Validators.validateExists(movieEntity, MOVIE_TO_ARGUMENT);
        for (final GenreTO genre : movie.getGenres()) {
            Validators.validateExists(genreService.get(genre.getId()), GENRE_TO_ARGUMENT);
        }

        final Movie updatedMovie = converter.convert(movie, Movie.class);
        updatedMovie.setMedia(getUpdatedMedia(movieEntity.getMedia(), movie.getMedia()));

        movieService.update(updatedMovie);
    }

    /**
     * @throws IllegalArgumentException                                  {@inheritDoc}
     * @throws cz.vhromada.validators.exceptions.ValidationException     {@inheritDoc}
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException {@inheritDoc}
     */
    @Override
    public void remove(final MovieTO movie) {
        movieTOValidator.validateMovieTOWithId(movie);
        final Movie movieEntity = movieService.get(movie.getId());
        Validators.validateExists(movieEntity, MOVIE_TO_ARGUMENT);

        movieService.remove(movieEntity);
    }

    /**
     * @throws IllegalArgumentException                                  {@inheritDoc}
     * @throws cz.vhromada.validators.exceptions.ValidationException     {@inheritDoc}
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException {@inheritDoc}
     */
    @Override
    public void duplicate(final MovieTO movie) {
        movieTOValidator.validateMovieTOWithId(movie);
        final Movie movieEntity = movieService.get(movie.getId());
        Validators.validateExists(movieEntity, MOVIE_TO_ARGUMENT);

        movieService.duplicate(movieEntity);
    }

    /**
     * @throws IllegalArgumentException                                  {@inheritDoc}
     * @throws cz.vhromada.validators.exceptions.ValidationException     {@inheritDoc}
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException {@inheritDoc}
     */
    @Override
    public void moveUp(final MovieTO movie) {
        movieTOValidator.validateMovieTOWithId(movie);
        final Movie movieEntity = movieService.get(movie.getId());
        Validators.validateExists(movieEntity, MOVIE_TO_ARGUMENT);
        final List<Movie> movies = movieService.getAll();
        Validators.validateMoveUp(movies, movieEntity, MOVIE_TO_ARGUMENT);

        movieService.moveUp(movieEntity);
    }

    /**
     * @throws IllegalArgumentException                                  {@inheritDoc}
     * @throws cz.vhromada.validators.exceptions.ValidationException     {@inheritDoc}
     * @throws cz.vhromada.validators.exceptions.RecordNotFoundException {@inheritDoc}
     */
    @Override
    public void moveDown(final MovieTO movie) {
        movieTOValidator.validateMovieTOWithId(movie);
        final Movie movieEntity = movieService.get(movie.getId());
        Validators.validateExists(movieEntity, MOVIE_TO_ARGUMENT);
        final List<Movie> movies = movieService.getAll();
        Validators.validateMoveDown(movies, movieEntity, MOVIE_TO_ARGUMENT);

        movieService.moveDown(movieEntity);
    }

    @Override
    public void updatePositions() {
        movieService.updatePositions();
    }

    @Override
    public int getTotalMediaCount() {
        int totalMediaCount = 0;
        for (final Movie movie : movieService.getAll()) {
            totalMediaCount += movie.getMedia().size();
        }

        return totalMediaCount;
    }

    @Override
    public Time getTotalLength() {
        int totalLength = 0;
        for (final Movie movie : movieService.getAll()) {
            for (final Medium medium : movie.getMedia()) {
                totalLength += medium.getLength();
            }
        }

        return new Time(totalLength);
    }

    /**
     * Updates media.
     *
     * @param originalMedia original list of media
     * @param updatedMedia  updated list of TO for medium
     * @return updated media
     */
    private static List<Medium> getUpdatedMedia(final List<Medium> originalMedia, final List<MediumTO> updatedMedia) {
        final List<Medium> result = new ArrayList<>();

        int index = 0;
        final int max = Math.min(originalMedia.size(), updatedMedia.size());
        while (index < max) {
            final Medium medium = new Medium();
            medium.setId(originalMedia.get(index).getId());
            medium.setNumber(index + 1);
            medium.setLength(updatedMedia.get(index).getLength());
            result.add(medium);
            index++;
        }
        while (index < updatedMedia.size()) {
            final Medium medium = new Medium();
            medium.setNumber(index + 1);
            medium.setLength(updatedMedia.get(index).getLength());
            result.add(medium);
            index++;
        }

        return result;
    }

}

File: catalog-core/src/main/kotlin/com/github/vhromada/catalog/facade/impl/MovieFacadeImpl.kt
package com.github.vhromada.catalog.facade.impl

import com.github.vhromada.catalog.entity.Genre
import com.github.vhromada.catalog.entity.Medium
import com.github.vhromada.catalog.entity.Movie
import com.github.vhromada.catalog.entity.Picture
import com.github.vhromada.catalog.facade.MovieFacade
import com.github.vhromada.common.entity.Time
import com.github.vhromada.common.facade.AbstractParentFacade
import com.github.vhromada.common.mapper.Mapper
import com.github.vhromada.common.result.Result
import com.github.vhromada.common.service.ParentService
import com.github.vhromada.common.validator.Validator
import org.springframework.stereotype.Component
import kotlin.math.min

/**
 * A class represents implementation of facade for movies.
 *
 * @author Vladimir Hromada
 */
@Component("movieFacade")
class MovieFacadeImpl(
    private val movieService: ParentService<com.github.vhromada.catalog.domain.Movie>,
    private val pictureService: ParentService<com.github.vhromada.catalog.domain.Picture>,
    private val genreService: ParentService<com.github.vhromada.catalog.domain.Genre>,
    mapper: Mapper<Movie, com.github.vhromada.catalog.domain.Movie>,
    movieValidator: Validator<Movie, com.github.vhromada.catalog.domain.Movie>,
    private val pictureValidator: Validator<Picture, com.github.vhromada.catalog.domain.Picture>,
    private val genreValidator: Validator<Genre, com.github.vhromada.catalog.domain.Genre>
) : AbstractParentFacade<Movie, com.github.vhromada.catalog.domain.Movie>(parentService = movieService, mapper = mapper, validator = movieValidator), MovieFacade {

    override fun updateData(data: Movie): Result<Unit> {
        val storedMovie = movieService.get(id = data.id!!)
        val validationResult = validator.validateExists(data = storedMovie)
        val pictureValidationResult = getPicture(movie = data)
        val genresValidationResult = getGenres(movie = data)
        if (validationResult.isOk() && pictureValidationResult.isOk() && genresValidationResult.isOk()) {
            val movie = mapper.map(source = data)
                .copy(media = getUpdatedMedia(originalMedia = storedMovie.get().media, updatedMedia = data.media!!))
            movie.createdUser = storedMovie.get().createdUser
            movie.createdTime = storedMovie.get().createdTime
            movieService.update(data = movie)
        }
        return Result.of(validationResult, pictureValidationResult, genresValidationResult)
    }

    @Suppress("DuplicatedCode")
    override fun addData(data: Movie): Result<Unit> {
        val pictureValidationResult = getPicture(movie = data)
        val genresValidationResult = getGenres(movie = data)
        if (pictureValidationResult.isOk() && genresValidationResult.isOk()) {
            val movie = mapper.map(source = data)
                .copy(genres = genresValidationResult.data!!)
            movieService.add(data = movie)
        }
        return Result.of(pictureValidationResult, genresValidationResult)
    }

    override fun getTotalMediaCount(): Result<Int> {
        return Result.of(data = movieService.getAll().sumBy { it.media.size })
    }

    override fun getTotalLength(): Result<Time> {
        return Result.of(data = Time(length = movieService.getAll().sumBy { it.media.sumBy { m -> m.length } }))
    }

    /**
     * Returns result with picture and validation errors.
     *
     * @param movie movie
     * @returns result with picture and validation errors
     */
    @Suppress("DuplicatedCode")
    private fun getPicture(movie: Movie): Result<com.github.vhromada.catalog.domain.Picture> {
        if (movie.picture == null) {
            return Result()
        }
        val picture = pictureService.get(id = movie.picture)
        val validationResult = pictureValidator.validateExists(data = picture)
        return if (validationResult.isOk()) {
            Result.of(data = picture.get())
        } else {
            Result.of(validationResult)
        }
    }

    /**
     * Returns result with genres and validation errors.
     *
     * @param movie movie
     * @returns result with genres and validation errors
     */
    @Suppress("DuplicatedCode")
    private fun getGenres(movie: Movie): Result<MutableList<com.github.vhromada.catalog.domain.Genre>> {
        val genres = mutableListOf<com.github.vhromada.catalog.domain.Genre>()
        val result = Result.of(genres)
        movie.genres!!.filterNotNull()
            .map { genreService.get(id = it.id!!) }
            .forEach {
                val validationResult = genreValidator.validateExists(data = it)
                if (validationResult.isOk()) {
                    genres.add(it.get())
                } else {
                    result.addEvents(eventList = validationResult.events())
                }
            }
        return result
    }

    /**
     * Updates media.
     *
     * @param originalMedia original media
     * @param updatedMedia  updated media
     * @return updated media
     */
    private fun getUpdatedMedia(originalMedia: List<com.github.vhromada.catalog.domain.Medium>, updatedMedia: List<Medium?>): List<com.github.vhromada.catalog.domain.Medium> {
        val result = mutableListOf<com.github.vhromada.catalog.domain.Medium>()

        var index = 0
        val max = min(originalMedia.size, updatedMedia.size)
        while (index < max) {
            val originalMedium = originalMedia[index]
            val updatedMedium = updatedMedia[index]!!
            val medium = com.github.vhromada.catalog.domain.Medium(id = originalMedium.id, number = index + 1, length = updatedMedium.length!!)
            medium.createdUser = originalMedium.createdUser
            medium.createdTime = originalMedium.createdTime
            result.add(medium)
            index++
        }
        while (index < updatedMedia.size) {
            val medium = com.github.vhromada.catalog.domain.Medium(id = null, number = index + 1, length = updatedMedia[index]!!.length!!)
            result.add(medium)
            index++
        }

        return result
    }

}
